common/lockmanager_test.go

```diff
@@ -0,0 +1,130 @@
+package common
+
+import (
+    "sync"
+    "testing"
+    "time"
+
+    "github.com/edulinq/autograder/config"
+)
+
+var (
+    key1 = "testkey1";
+    key2 = "testkey2";
+    doesNotExistKey = "dne";
+    staleDuration = time.Duration(config.STALELOCK_DURATION_SECS.Get()) * time.Second;
+    lockWaitGroup sync.WaitGroup;
+)
+
+func TestLockBase(t *testing.T) {
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        t.Errorf("Failed to unlock a key.");
+    }
+}
+
+func TestUnlockingAnUnlockedLock(t *testing.T) {
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        t.Errorf("Failed to unlock a key.");
+    }
+
+    err = Unlock(key1);
+    if (err == nil) {
+        t.Errorf("Unlocking a lock that is already unlocked did not return an error.");
+    }
+}
+
+func TestUnlockingKeyThatIsntLocked(t *testing.T) {
+    err := Unlock(doesNotExistKey);
+    if (err == nil) {
+        t.Errorf("Lock manager unlocked a key that wasn't locked.");
+    }
+}
+
+func TestConcurrentLockingUnlocking(t *testing.T) {
+    // Lock key1 and key2 for the first time.
+    Lock(key1);
+    Lock(key2);
+
+    lockWaitGroup.Add(2);
+
+    // This goroutine should wait to lock/unlock key1 a second time until it gets
+    // unlocked for the first time.
+    go func() {
+        defer lockWaitGroup.Done();
+        Lock(key1);
+        defer func() {
+            err := Unlock(key1);
+            if (err != nil) {
+                t.Errorf("Failed to unlock a key.");
+            }
+        }()
+    }()
+
+    // This goroutine should wait to lock/unlock key2 a second time until it gets
+    // unlocked for the first time.
+    go func() {
+        defer lockWaitGroup.Done();
+        Lock(key2);
+        defer func() {
+            err := Unlock(key2);
+            if (err != nil) {
+                t.Errorf("Failed to unlock a key.");
+            }
+        }()
+    }()
+    
```

Comment:
You need to verify that the other thread is still waiting before you unlock.
(Otherwise, you don't know if you hit the use case.)

This is actually a use case for a WaitGroup, but instead of being between the locking threads, it would be between the locking and main threads.

Reply:
Would it make more sense to use a channel as well? For example, the go routine could send True to the channel right before it tries to acquire the lock, and after the go func there could be a small sleep (or a waitgroup waiting) and then an if statement to confirm True got sent to the channel which indicates the go routine is waiting. I might be failing to see how just a waitgroup would be able to trigger a test fail.

Reply:
You can use either.

The idea here is just that you want to wait until both threads hit a specific location to continue.
You can do this by blocking on a channel or by blocking on a wait group.
You just need your threads to block so you can ensure they both are at the correct spot.
Note that if you used a channel, you would have to be a bit careful since you don't know which thread will execute first (after the go routine is spawned).